### ISSUE TITLE
multiple library versions and transitive dependency leaking

### ISSUE THREAD

**Issue by nlacombe42** [nlacombe42/prophecy#8]
I will want to find a solution to those 2 dependency issues:
1. multiple versions of the same library at runtime
  - ex. B1 means library B version 1
    - app A depends on B1 and C1
    - B1 depends on D1 (and potentially exposes D1's types)
    - C1 depends on D2 (and potentially exposes D2's types)
    - D1 and D2's contract or implementation might be incompatible: B1 might not be able to work with D2 and C1 might not be able to work with D1
  - in such scenarios D1's types even if they have the same name and interface, might not be compatible with D2's types because there might be implementation differences (different private fields or algorithm/contract ot different format in which it stores its state)
  - also, some library might want to share state with all other version of the same library in the same process/runtime
    most of the time this is a bad idea, but in some cases like the following it is a good idea
    ex: the library creates another process or os resource that is not managed by the pl runtime, and the library creates a pl object (class instance) to represent this resource inside the pl runtime: so that when it is no longer needed it is killed/freed or so that this resource be shared by all consumer of the library on the runtime (if the resource can only be craeted once in the os for instance)
    having multiple runtime object for that resource might make it possible to try dangerous or not wnated operations such as trying to free the same resource twice or trying to create twice, etc
2. transitive dependency leaking
  - ex
    - A depends on B
    - B depends on C
  - in that case you want to make sure that if A uses a class from C that it declares C as a dependency
  - also, B might want to prevent any consumer (such as A) to use C since it's an implementation dependency that they might want to swap out in the future

to fix issue 2 I would like the pl to include all of those tools: compiler, package manager, task runner (for build and more, à la gradle)
that way the user could specify the version, name, dependencies and build tasks inside the same file, similalry as gradle
and the compiler could use that information to validate and optimize certain things
for instance, when specifying dependencies the user could specify if the dependency is to be exposed or not to consumers and which part of its code is exposed
something that would be similar to a merge between java's build.gradle and module-info.java
